File: Programmers/60059.js
// 문제를 정말 잘 읽어야 한다. 
// 자물쇠의 돌기와 열쇠의 돌기가 만나면 안된다는 조건을 간과하여 오래걸렸다. 
function canOpen(key, needKey, dr, dc, numLock) {
    const keyRow = key.length;
    const keyCol = key[0].length;
    const needKeyRow = needKey.length;
    const needKeyCol = needKey[0].length;
    let sum = 0;
    let nr, nc;
    for(let r = 0; r < keyRow; r++){
        for(let c = 0; c < keyCol; c++){
            nr = r + dr;
            nc = c + dc;
            if(nr < 0 || nr >= needKeyRow || nc < 0 || nc >= needKeyCol){
                continue;
            }
            if(key[r][c] === 1 && needKey[nr][nc] === 0){
                return false;
            }
            sum += (key[r][c] === 1 && needKey[nr][nc] === 1) ? 1 : 0;
        }
    }
    return sum === numLock;
}


function rotate90(key){
    const row = key.length;
    const col = key[0].length
    const result = Array.from(Array(row), () => Array(col).fill(0));
    for(let r = 0; r < row; r++){
        for(let c = 0; c < col; c++){
            result[r][c] = key[c][row-r-1];
        }
    }
    return result;
}


function solution(key, lock) {
    const needKey = lock.map(row => {
        return row.map(item => 1-item)
    });
    const numLock = needKey.reduce((prev, curr) => {
        return prev + curr.reduce((p, c) => {
            return p + c;
        }, 0)
    }, 0);
    const numKey = key.reduce((prev, curr) => {
        return prev + curr.reduce((p, c) => {
            return p + c;
        }, 0)
    }, 0)

    if(numLock > numKey){
        return false;
    }
    // key sliding window
    const row = needKey.length;
    const col = needKey[0].length;
    const keyRow = key.length;
    const keyCol = key[0].length;
    let currKey = key;
    for(let i = 0; i < 4; i++){
        currKey = rotate90(currKey);   
        for(let dr = -keyRow + 1; dr < row; dr++){
            for(let dc = -keyCol + 1; dc < col; dc++){
                if(canOpen(currKey, needKey, dr, dc, numLock)){
                    return true;
                }
            }
        }
    }
    return false;
}

File: Programmers/60059.py
def rotate(key):
    row, col = len(key), len(key[0])
    rotate_key = [[0] * col for _ in range(row)]
    for r in range(row):
        for c in range(col):
            rotate_key[r][c] = key[c][row-1-r]
    return rotate_key

def canOpen(key, need_key, dr, dc, num_lock):
    key_row, key_col = len(key), len(key[0])
    lock_row, lock_col = len(need_key), len(need_key[0])
    add = 0
    for r in range(key_row):
        for c in range(key_col):
            nr, nc = r + dr, c + dc
            if not(0 <= nr < lock_row and 0 <= nc < lock_col):
                continue
            if key[r][c] == 1 and need_key[nr][nc] == 0:
                return False
            add += 1 if key[r][c] == 1 and need_key[nr][nc] == 1 else 0
    return add == num_lock
            
def solution(key, lock):
    num_key, num_lock = 0, 0
    need_key = [[0] * len(lock[0]) for _ in range(len(lock))]
    key_row, key_col = len(key), len(key[0])
    lock_row, lock_col = len(lock), len(lock[0])
    for i in range(key_row):
        for j in range(key_col):
            num_key += key[i][j]
    for i in range(lock_row):
        for j in range(lock_col):
            need_key[i][j] = 1 - lock[i][j]
            num_lock += need_key[i][j]
            
    if num_key < num_lock:
        return False
    
    # sliding window
    curr_key = key
    for _ in range(4):
        curr_key = rotate(curr_key)
        for dr in range(-key_row + 1, lock_row):
            for dc in range(-key_col + 1, lock_col):
                if canOpen(curr_key, need_key, dr, dc, num_lock):
                    return True
    return False